File: app/src/main/java/com/elementary/tasks/core/file_explorer/FileExplorerActivity.java
package com.elementary.tasks.core.file_explorer;

import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.elementary.tasks.R;
import com.elementary.tasks.core.ThemedActivity;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Dialogues;
import com.elementary.tasks.core.utils.MeasureUtils;
import com.elementary.tasks.core.utils.Permissions;
import com.elementary.tasks.core.utils.Sound;
import com.elementary.tasks.core.utils.ViewUtils;
import com.elementary.tasks.core.views.roboto.RoboEditText;
import com.elementary.tasks.core.views.roboto.RoboTextView;
import com.elementary.tasks.databinding.ActivityFileExplorerBinding;
import com.squareup.picasso.Picasso;

import java.io.File;
import java.io.FilenameFilter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class FileExplorerActivity extends ThemedActivity {

    private static final String TAG = "FileExplorerActivity";
    private static final int SD_CARD = 444;
    public static final String TYPE_MUSIC = "music";
    public static final String TYPE_PHOTO = "photo";

    private ArrayList<String> str = new ArrayList<>();
    private Boolean firstLvl = true;
    private boolean isDark = false;
    private boolean mFilter;

    private ArrayList<FileDataItem> mDataList;
    private File path = new File(Environment.getExternalStorageDirectory() + "");
    private String mFileName;
    private String mFilePath;

    private String filType;

    private FileRecyclerAdapter mAdapter;
    private Sound mSound;

    private ActivityFileExplorerBinding binding;
    private RecyclerView mFilesList;
    private LinearLayout mPlayerLayout;
    private RoboTextView mMelodyTitle;
    private RoboEditText mSearchView;

    private RecyclerClickListener recyclerClick = this::selectFile;

    private FilterCallback mFilterCallback = new FilterCallback() {
        @Override
        public void filter(int size) {
            mFilesList.scrollToPosition(0);
        }
    };

    private void selectFile(int position) {
        FileDataItem item = mAdapter.getItem(position);
        mFileName = item.getFileName();
        mFilePath = item.getFilePath();
        File sel = new File(path + "/" + mFileName);
        if (sel.isDirectory()) {
            firstLvl = false;
            str.add(mFileName);
            mDataList = null;
            path = new File(sel + "");
            loadFileList();
            loadList();
        } else if (mFileName.equalsIgnoreCase(getString(R.string.up)) && !sel.exists()) {
            moveUp();
        } else {
            if (filType.matches("any")) {
                sendFile();
            } else if (filType.equals(TYPE_PHOTO)) {
                if (isImage(mFileName)) {
                    showFullImage();
                } else {
                    Toast.makeText(this, "Not a image file", Toast.LENGTH_SHORT).show();
                }
            } else {
                if (isMelody(mFileName)) {
                    play();
                } else {
                    Toast.makeText(this, getString(R.string.not_music_file), Toast.LENGTH_SHORT).show();
                }
            }
        }
    }

    private void showFullImage() {
        AlertDialog.Builder builder = Dialogues.getDialog(this);
        builder.setTitle(mFileName);
        ImageView imageView = new ImageView(this);
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, MeasureUtils.dp2px(this, 256));
        imageView.setLayoutParams(layoutParams);
        Picasso.with(this)
                .load(new File(mFilePath))
                .into(imageView);
        builder.setView(imageView);
        builder.setPositiveButton(getString(R.string.ok), (dialog, which) -> {
            dialog.dismiss();
            sendFile();
        });
        builder.setNegativeButton(getString(R.string.cancel), (dialog, which) -> dialog.dismiss());
        builder.create().show();
    }

    private void moveUp() {
        String s = str.remove(str.size() - 1);
        path = new File(path.toString().substring(0,
                path.toString().lastIndexOf(s)));
        mDataList = null;
        if (str.isEmpty()) {
            firstLvl = true;
        }
        loadFileList();
        loadList();
    }

    private void sendFile() {
        Intent intent = new Intent();
        intent.putExtra(Constants.FILE_PICKED, mFilePath);
        setResult(RESULT_OK, intent);
        finish();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mSound = new Sound(this);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_file_explorer);
        filType = getIntent().getStringExtra(Constants.FILE_TYPE);
        if (filType == null) filType = TYPE_MUSIC;
        isDark = getThemeUtil().isDark();
        initActionBar();
        initRecyclerView();
        initPlayer();
        initSearch();
        initButtons();
        if (Permissions.checkPermission(this, Permissions.READ_EXTERNAL)) {
            loadFileList();
            loadList();
        } else {
            Permissions.requestPermission(this, SD_CARD, Permissions.READ_EXTERNAL);
        }
    }

    private void initPlayer() {
        mPlayerLayout = binding.playerLayout;
        mPlayerLayout.setVisibility(View.GONE);
        mMelodyTitle = binding.currentMelody;
    }

    private void initRecyclerView() {
        mFilesList = binding.mDataList;
        mFilesList.setHasFixedSize(true);
        mFilesList.setLayoutManager(new LinearLayoutManager(this));
    }

    private void initActionBar() {
        Toolbar toolbar = binding.toolbar;
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
    }

    private void initSearch() {
        mSearchView = binding.searchField;
        mSearchView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (mFilter && mAdapter != null) mAdapter.filter(s.toString(), mDataList);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void initButtons() {
        ImageButton clearButton = binding.clearButton;
        binding.fab.setOnClickListener(mListener);
        binding.pauseButton.setOnClickListener(mListener);
        binding.stopButton.setOnClickListener(mListener);
        binding.playButton.setOnClickListener(mListener);
        clearButton.setOnClickListener(mListener);
    }

    private void loadList() {
        if (mDataList == null) {
            Toast.makeText(this, getString(R.string.no_files), Toast.LENGTH_SHORT).show();
            finish();
        }
        mFilesList.setAdapter(mAdapter);
    }

    private void play() {
        if (!mSound.isPlaying()) {
            if (mPlayerLayout.getVisibility() == View.GONE) {
                ViewUtils.expand(mPlayerLayout);
            }
            if (mSound.isPaused() && mSound.isSameFile(mFilePath)) {
                mSound.resume();
            } else {
                mSound.play(mFilePath);
                mMelodyTitle.setText(mFileName);
            }
        } else {
            if (mSound.isSameFile(mFilePath)) {
                return;
            }
            mSound.play(mFilePath);
            mMelodyTitle.setText(mFileName);
        }
    }

    private void pause() {
        if (mSound.isPlaying()) {
            mSound.pause();
        }
    }

    private void stop() {
        if (mSound.isPlaying()) {
            mSound.stop(true);
        }
        ViewUtils.collapse(mPlayerLayout);
    }

    private void loadFileList() {
        try {
            path.mkdirs();
        } catch (SecurityException e) {
            e.printStackTrace();
        }
        mFilter = false;
        mSearchView.setText("");
        mFilter = true;
        if (path.exists()) {
            createFilteredFileList();
        }
        mAdapter = new FileRecyclerAdapter(this, mDataList, recyclerClick, mFilterCallback);
    }

    private void createFilteredFileList() {
        FilenameFilter filter = (dir, filename) -> {
            File sel = new File(dir, filename);
            return (sel.isFile() || sel.isDirectory())
                    && !sel.isHidden();
        };

        List<String> list;
        try {
            list = Arrays.asList(path.list(filter));
        } catch (NullPointerException e) {
            list = new ArrayList<>();
        }
        Collections.sort(list);
        mDataList = new ArrayList<>(list.size());
        for (int i = 0; i < list.size(); i++) {
            String fileName = list.get(i);
            File sel = new File(path, fileName);
            mDataList.add(i, new FileDataItem(fileName, 0, sel.toString()));

            if (sel.isDirectory()) {
                mDataList.get(i).setIcon(getDirectoryIcon());
            }
        }

        if (!firstLvl) {
            addUpItem();
        }
    }

    private void addUpItem() {
        ArrayList<FileDataItem> temp = new ArrayList<>(mDataList.size() + 1);
        temp.add(0, new FileDataItem(getString(R.string.up), getUndoIcon(), ""));
        temp.addAll(mDataList);
        mDataList = temp;
    }

    private boolean isMelody(String file) {
        return file != null && (file.endsWith(".mp3") || file.endsWith(".ogg")
                || file.endsWith(".m4a") || file.endsWith(".flac"));
    }

    private boolean isImage(String file) {
        return file != null && (file.endsWith(".jpg") || file.endsWith(".jpeg")
                || file.endsWith(".png") || file.endsWith(".tiff"));
    }

    private int getDirectoryIcon() {
        return isDark ? R.drawable.ic_folder_white_24dp : R.drawable.ic_folder_black_24dp;
    }

    private int getUndoIcon() {
        return isDark ? R.drawable.ic_undo_white_24dp : R.drawable.ic_undo_black_24dp;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                exit();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onBackPressed() {
        if (!firstLvl) {
            moveUp();
        } else {
            exit();
        }
    }

    private void exit() {
        if (isMelody(mFileName)) {
            stop();
        }
        setResult(RESULT_CANCELED);
        finish();
    }

    private View.OnClickListener mListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            switch (v.getId()) {
                case R.id.fab:
                    saveChoice();
                    break;
                case R.id.playButton:
                    play();
                    break;
                case R.id.stopButton:
                    stop();
                    break;
                case R.id.pauseButton:
                    pause();
                    break;
                case R.id.clearButton:
                    mSearchView.setText("");
                    break;
            }
        }
    };

    private void saveChoice() {
        if (isMelody(mFileName)) {
            stop();
            sendFile();
        } else {
            Toast.makeText(this, getString(R.string.not_music_file), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case SD_CARD:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    loadFileList();
                    loadList();
                } else {
                    setResult(RESULT_CANCELED);
                    finish();
                }
                break;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/file_explorer/FileExplorerActivity.kt
package com.elementary.tasks.core.file_explorer

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.elementary.tasks.R
import com.elementary.tasks.core.BindingActivity
import com.elementary.tasks.core.filter.SearchModifier
import com.elementary.tasks.core.interfaces.ActionsListener
import com.elementary.tasks.core.utils.*
import com.elementary.tasks.databinding.ActivityFileExplorerBinding
import com.elementary.tasks.databinding.ListItemImageBinding
import com.squareup.picasso.Picasso
import timber.log.Timber
import java.io.File

class FileExplorerActivity : BindingActivity<ActivityFileExplorerBinding>(R.layout.activity_file_explorer) {

    private lateinit var viewModel: SelectFileViewModel
    private var mFileName: String = ""
    private var mFilePath: String = ""

    private val mAdapter: FileRecyclerAdapter = FileRecyclerAdapter()
    private val searchModifier = object : SearchModifier<FileItem>(null, {
        mAdapter.submitList(it)
        showList(it.size)
    }) {
        override fun filter(v: FileItem): Boolean {
            return searchValue.isEmpty() || v.fileName.toLowerCase().contains(searchValue.toLowerCase())
        }
    }
    private var mSound: Sound? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(this).get(SelectFileViewModel::class.java)
        viewModel.files.observe(this, androidx.lifecycle.Observer { files ->
            files?.let {
                searchModifier.original = it
            }
        })
        viewModel.isLoading.observe(this, androidx.lifecycle.Observer { isLoading ->
            isLoading?.let {
                if (it) binding.loaderView.visibility = View.VISIBLE
                else binding.loaderView.visibility = View.GONE
            }
        })

        mSound = Sound(this, prefs)

        viewModel.filType = intent.getStringExtra(Constants.FILE_TYPE) ?: ""
        if (viewModel.filType == "") viewModel.filType = TYPE_MUSIC

        initRecyclerView()
        initPlayer()
        initSearch()
        initButtons()
    }

    override fun onStart() {
        super.onStart()
        if (Permissions.ensurePermissions(this, SD_CARD, Permissions.READ_EXTERNAL)) {
            viewModel.loadFileList()
        }
    }

    private fun longClick(position: Int) {
        val item = mAdapter.getFileItem(position) ?: return
        mFileName = item.fileName
        mFilePath = item.filePath
        if (isImage(item.fileName)) {
            showFullImage()
        }
    }

    private fun selectFile(position: Int) {
        val item = mAdapter.getFileItem(position) ?: return
        Timber.d("selectFile: $item")
        mFileName = item.fileName
        mFilePath = item.filePath
        val sel = File(item.filePath)
        if (sel.isDirectory) {
            viewModel.str.add(mFileName)
            binding.searchField.setText("")
            viewModel.loadFileList(sel, false)
        } else if (item.isUp) {
            moveUp()
        } else {
            if (viewModel.filType.matches("any".toRegex())) {
                sendFile()
            } else if (viewModel.filType == TYPE_PHOTO) {
                if (isImage(mFileName)) {
                    showFullImage()
                } else {
                    Toast.makeText(this, R.string.not_a_image_file, Toast.LENGTH_SHORT).show()
                }
            } else {
                if (isMelody(mFileName)) {
                    play()
                } else {
                    Toast.makeText(this, getString(R.string.not_music_file), Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun showFullImage() {
        val builder = dialogues.getMaterialDialog(this)
        builder.setTitle(mFileName)
        val imageView = ListItemImageBinding.inflate(LayoutInflater.from(this))
        Picasso.get()
                .load(File(mFilePath))
                .resize(512, 512)
                .centerCrop()
                .into(imageView.photoView)
        builder.setView(imageView.root)
        builder.setPositiveButton(getString(R.string.ok)) { dialog, _ ->
            dialog.dismiss()
            sendFile()
        }
        builder.setNegativeButton(getString(R.string.cancel)) { dialog, _ -> dialog.dismiss() }
        builder.create().show()
    }

    private fun moveUp() {
        val index = viewModel.str.size - 1
        if (index != -1) {
            val s = viewModel.str.removeAt(index)
            viewModel.moveUp(s, viewModel.str.isEmpty())
        }
    }

    private fun sendFile() {
        val intent = Intent()
        intent.putExtra(Constants.FILE_PICKED, mFilePath)
        setResult(RESULT_OK, intent)
        finish()
    }

    private fun initPlayer() {
        binding.playerLayout.visibility = View.GONE
    }

    private fun initRecyclerView() {
        mAdapter.clickListener = object : ActionsListener<FileItem> {
            override fun onAction(view: View, position: Int, t: FileItem?, actions: ListActions) {
                when (actions) {
                    ListActions.MORE -> longClick(position)
                    ListActions.OPEN -> selectFile(position)
                    else -> {
                    }
                }
            }
        }
        if (prefs.isTwoColsEnabled && ViewUtils.isHorizontal(this)) {
            binding.recyclerView.layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        } else {
            binding.recyclerView.layoutManager = LinearLayoutManager(this)
        }
        binding.recyclerView.adapter = mAdapter
        binding.recyclerView.isNestedScrollingEnabled = false
        ViewUtils.listenScrollableView(binding.scroller) {
            binding.toolbarView.isSelected = it > 0
        }
    }

    private fun initSearch() {
        binding.searchField.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                searchModifier.setSearchValue(s.toString())
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
    }

    private fun initButtons() {
        binding.backButton.setOnClickListener { onBackPressed() }
        binding.saveButton.setOnClickListener { saveChoice() }
        binding.pauseButton.setOnClickListener { pause() }
        binding.stopButton.setOnClickListener { stop() }
        binding.playButton.setOnClickListener { play() }
    }

    private fun showList(count: Int) {
        if (count == 0) {
            Toast.makeText(this, getString(R.string.no_files), Toast.LENGTH_SHORT).show()
        }
        binding.recyclerView.smoothScrollToPosition(0)
        refreshView(count)
    }

    private fun play() {
        if (mSound?.isPlaying == false) {
            if (binding.playerLayout.visibility == View.GONE) {
                binding.playerLayout.visibility = View.VISIBLE
            }
            if (mSound?.isPaused == true && mSound?.isSameFile(mFilePath) == true) {
                mSound?.resume()
            } else {
                mSound?.play(mFilePath)
                binding.currentMelody.text = mFileName
            }
        } else {
            if (mSound?.isSameFile(mFilePath) == true) {
                return
            }
            mSound?.play(mFilePath)
            binding.currentMelody.text = mFileName
        }
    }

    private fun pause() {
        if (mSound?.isPlaying == true) {
            mSound?.pause()
        }
    }

    private fun stop() {
        if (mSound?.isPlaying == true) {
            mSound?.stop(true)
        }
        binding.playerLayout.visibility = View.GONE
    }

    private fun isMelody(file: String?): Boolean {
        return file != null && (file.endsWith(".mp3") || file.endsWith(".ogg")
                || file.endsWith(".m4a") || file.endsWith(".flac"))
    }

    private fun isImage(file: String?): Boolean {
        return file != null && (file.endsWith(".jpg") || file.endsWith(".jpeg")
                || file.endsWith(".png") || file.endsWith(".tiff"))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                exit()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onBackPressed() {
        if ((!viewModel.isRoot)) {
            moveUp()
        } else {
            exit()
        }
    }

    private fun exit() {
        if (isMelody(mFileName)) {
            stop()
        }
        setResult(Activity.RESULT_CANCELED)
        finish()
    }

    private fun saveChoice() {
        if (isMelody(mFileName)) {
            stop()
            sendFile()
        } else {
            Toast.makeText(this, getString(R.string.not_music_file), Toast.LENGTH_SHORT).show()
        }
    }

    private fun refreshView(count: Int) {
        if (count > 0) {
            binding.emptyItem.visibility = View.GONE
            binding.scroller.visibility = View.VISIBLE
        } else {
            binding.scroller.visibility = View.GONE
            binding.emptyItem.visibility = View.VISIBLE
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            SD_CARD -> if (Permissions.isAllGranted(grantResults)) {
                viewModel.loadFileList()
            } else {
                setResult(Activity.RESULT_CANCELED)
                finish()
            }
        }
    }

    companion object {
        private const val SD_CARD = 444
        const val TYPE_MUSIC = "music"
        const val TYPE_PHOTO = "photo"
    }
}
